Add Varnish

kierenevans opened the pull request:
If varnish is in the list of services:
* Use `varnish` as the entrypoint from traefik.
* Use `tls-offload` as the target for acceptance testing (behat) within the docker-compose environment to provide TLS support.
* Deploy `varnish` statefulset to pipeline environments as the entrypoint from the ingress.
* Deploy a configmap containing the varnish configuration, to avoid wiping out the cache on each deployment if built with docker image+configuration per-commit.
* Configure the "timeout chain" of varnish and nginx to be the max execution time of php plus one or two seconds. The idea is to never show a 502/504 when php is still working, though this may not be true if PHP is performing slow network requests as they block and don't count towards max_execution_time, or if the queue to access php-fpm from nginx is high.
* BREAKING CHANGE: `.Values.ingress` renamed to `.Values.ingress.type` to allow for more ingress configurations.
* Magento 2 VCL from https://github.com/magento/magento2/blob/2.3.5/app/code/Magento/PageCache/etc/varnish6.vcl used as the configuration.
* Magento 2 env.php configured for http_cache_hosts

Fixes https://github.com/inviqa/harness-base-php/issues/14

kierenevans committed:
Stop routing to NGINX from traefik if varnish is present

kierenevans committed:
Add optional varnish service

kierenevans committed:
Add varnish dockerfile

kierenevans committed:
Define default max_execution_time so we can pick it up in nginx and
varnish

kierenevans committed:
Set max_execution_time - varying from the default 30s where there was already variance

kierenevans committed:
Add default varnish configuration from docker image, customised for nginx

kierenevans committed:
Add customised varnish configuration for magento2 based on https://github.com/magento/magento2/blob/2.3.5/app/code/Magento/PageCache/etc/varnish6.vcl

kierenevans committed:
Let Kubernetes handle the magento2 healthcheck

kierenevans committed:
Test with varnish

kierenevans committed:
Convert from built image which would change per-commit, to something we
can drive with a configmap

kierenevans committed:
Render via twig

kierenevans committed:
Fix for type issue for varnish seconds

kierenevans committed:
Fix shellcheck issues in config_render.sh.twig

kierenevans committed:
Add TLS Offloading image for integration testing via varnish

kierenevans committed:
Change service target when using varnish

kierenevans committed:
Add helm deployment for varnish

kierenevans committed:
Bring in the varnish configuration used by docker as a configmap.

This setup allows us to update the configuration without losing the
cache

kierenevans committed:
Define .Values.services.varnish

kierenevans committed:
Fix using non-twig syntax in twig file

kierenevans committed:
Fix not being able to embed files with "../"

kierenevans committed:
Fix template being invalid

kierenevans committed:
Fix .twig file being used in helm chart

kierenevans committed:
Fix typo in ingress targetService

kierenevans committed:
BREAKING: Fix type error - move ingress: "standard/istio" to ingress.type

kierenevans committed:
Bring up multiple varnish's at once, we only use statefulset for
consistent naming

kierenevans committed:
Remove invalid `status: {}` - a StatefulSet needs status.replicas set

kierenevans committed:
Support multiple replicas for varnish. Configure magento2

kierenevans committed:
Fix replicas being missing from values

kierenevans committed:
Use yaml directly

kierenevans committed:
Set magento2 caching application

kierenevans committed:
Remove example header

kierenevans committed:
Use underscores

kierenevans committed:
`nginx` is not the service name in kubernetes

kierenevans committed:
Render with resourcePrefix'd webapp for helm

kierenevans committed:
Avoid needing to change all the variations of values file

kierenevans committed:
Quotes

kierenevans committed:
Can't embed .Values in a value already

kierenevans committed:
Fix for spryker

tkotosz commented on src/magento2/harness.yml:
```diff
@@ -130,6 +133,9 @@ attributes:
         - run "magento config:set 'web/secure/use_in_adminhtml' '1'"
         - run "magento config:set 'web/seo/use_rewrites'        '1'"
+        - if [[ "$HAS_VARNISH" == "true" ]]; then
+            run "magento config:set 'system/full_page_cache/caching_application' '2'";
```
isn't there a few other config to set as well in this case? (like host port maybe?)

kierenevans replied:
Nope, I think they are just for the VCL generator

tkotosz replied:
ah I see :+1:

tkotosz reviewed (dismissed):
generally looks good to me. @dcole-inviqa could you please have a look at this, just to be on the safe side since it is a breaking change?

kierenevans committed:
Fix too many restarts

Importing https://varnish-cache.org/docs/6.2/whats-new/upgrading-6.2.html#vcl
from https://github.com/magento/magento2/pull/28137
